src/EventTeachingService.cpp

```diff
@@ -11,259 +11,427 @@
+      return PROCESSED;
+
+    case OPC_RQEVN:
+      // 58 - request for number of stored events
+      // DEBUG_SERIAL << F("> RQEVN -- number of stored events for nn = ") << nn << endl;
 
-  case OPC_NNULN:
-    // received NNULN -- exit from learn mode
+      if (nn == module_config->nodeNum) 
+      {
+        // respond with 0x74 NUMEV
+        controller->sendMessageWithNN(OPC_NUMEV, module_config->numEvents());
+      }
 
-    if (nn == module_config->nodeNum)
-    {
-      inhibitLearn();
-    }
+      return PROCESSED;
 
-    return PROCESSED;
+    case OPC_NERD:
+      // 57 - request for all stored events
+      DEBUG_SERIAL << F("> NERD : request all stored events for nn = ") << nn << endl;
 
-  case OPC_RQEVN:
-    // received RQEVN -- request for number of stored events
-    // DEBUG_SERIAL << F("> RQEVN -- number of stored events for nn = ") << nn << endl;
+      if (nn == module_config->nodeNum) 
+      {
+        msg->len = 8;
+        msg->data[0] = OPC_ENRSP;     // response opcode
+        msg->data[1] = highByte(nn);  // my NN hi
+        msg->data[2] = lowByte(nn);   // my NN lo
 
-    if (nn == module_config->nodeNum)
-    {
-      // respond with 0x74 NUMEV
+        for (byte i = 0; i < module_config->EE_MAX_EVENTS; i++) 
+        {
+          if (module_config->getEvTableEntry(i) != 0) 
+          {
+            // it's a valid stored event
+            // read the event data from EEPROM
+            // construct and send a ENRSP message
+            module_config->readEvent(i, &msg->data[3]);
+            msg->data[7] = i;  // event table index
 
-      controller->sendMessageWithNN(OPC_NUMEV, module_config->numEvents());
-    }
+            DEBUG_SERIAL << F("> sending ENRSP reply for event index = ") << i << endl;
+            controller->sendMessage(msg);
+            delay(10);
 
-    return PROCESSED;
+          }  // valid stored ev
+        }    // loop each ev
+      }      // for me
 
-  case OPC_NERD:
-    // request for all stored events
-    // DEBUG_SERIAL << F("> NERD : request all stored events for nn = ") << nn << endl;
+      return PROCESSED;
 
-    if (nn == module_config->nodeNum)
-    {
-      msg->len = 8;
-      msg->data[0] = OPC_ENRSP;                       // response opcode
-      msg->data[1] = highByte(module_config->nodeNum);        // my NN hi
-      msg->data[2] = lowByte(module_config->nodeNum);         // my NN lo
+    case OPC_NENRD:
+      // 72 - request read stored event by event index
+      // DEBUG_SERIAL << F("> NERD : request all stored events for nn = ") << nn << endl;
 
-      for (byte i = 0; i < module_config->EE_MAX_EVENTS; i++)
+      if (nn == module_config->nodeNum) 
       {
-        if (module_config->getEvTableEntry(i) != 0) {
-          // it's a valid stored event
+        msg->len = 8;
+        msg->data[0] = OPC_ENRSP;     // response opcode
+        msg->data[1] = highByte(nn);  // my NN hi
+        msg->data[2] = lowByte(nn);   // my NN lo
```

Comment:
These four lines that set the message data should be moved down to line 181 where we are sending this message.
First of all, you only need these data set when we are sending the message in the else branch.
Secondly it is confusing that you are writing to the msg structure and then on line 169 you read from it.

Reply:
Good point.  It's done